File: frmMain.cs
/*
                $$$$$$$$\   $$$$$\  $$$$$$\         $$$$$$\  $$\      $$\  $$$$$$\  
                $$  _____|  \__$$ |$$  __$$\       $$  __$$\ $$$\    $$$ |$$  __$$\ 
                $$ |           $$ |$$ /  \__|      $$ /  \__|$$$$\  $$$$ |$$ /  \__|
                $$$$$\         $$ |$$ |            \$$$$$$\  $$\$$\$$ $$ |\$$$$$$\  
                $$  __|  $$\   $$ |$$ |             \____$$\ $$ \$$$  $$ | \____$$\ 
                $$ |     $$ |  $$ |$$ |  $$\       $$\   $$ |$$ |\$  /$$ |$$\   $$ |
                $$$$$$$$\\$$$$$$  |\$$$$$$  |      \$$$$$$  |$$ | \_/ $$ |\$$$$$$  |
                \________|\______/  \______/        \______/ \__|     \__| \______/ 

                                EJC Short Message Service v0.5

                    Having troubled making your own Send and Read Messages code for your 
            Project / System using your GSM Modem Device? Say no more. EJC SMS (EJC Short 
            Message Service) is a custom class that use AT Comamnd will let you to Send 
            and Read Messages at ease in which helps you to minimized your time developing 
            your system with SMS Feature.

            For more Information about AT Commands please refer to this site:
                https://www.developershome.com/sms/

                                                  Developed by: https://github.com/eksqtr
 */

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace SMS_Sender
{
    public partial class frmMain : Form
    {
        // Intantiate SMS as EJCSMS object and configure some of the EJCSMS Configurations.
        EJCSMS SMS = new EJCSMS(allow_long_msg: true, allow_empty_sms: false);
        private const string GSM_MODEM_PORT = "COM5";

        public frmMain()
        {
            InitializeComponent();
        }
        private void frmMain_Load(object sender, EventArgs e)
        {
            lblDateToday.Text = DateTime.Now.Date.ToString("MMMM dd, yyyy");
            SMS.InitSMS(GSM_MODEM_PORT); // Initialize the SMS.

            // Easy Peasy showing all the SMS from your Sim Card Inbox in just one line of code.
            gridViewInbox.DataSource = SMS.ReadAllMessages();
        }

        private void bttnSend_Click(object sender, EventArgs e)
        {
            if(string.IsNullOrEmpty(txtMessage.Text))
            {
                MessageBox.Show("Message Box is empty! Please input something.", "Message Box is Empty", MessageBoxButtons.OK, MessageBoxIcon.Error);
                txtMessage.Focus();
                return;
            }
            if(checkSendMultiple.CheckState != CheckState.Checked) // Single SMS
            {
                if (!SMS.IsValidRecipientNumber(txtPhoneNumber.Text))
                {
                    MessageBox.Show("Invalid Phone Number!", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    txtPhoneNumber.Focus();
                    return;
                }
                if (SMS.SendSMS(txtPhoneNumber.Text, txtMessage.Text))
                {
                    MessageBox.Show("Message has successfully sent!", "Delivering SMS Successfuly", MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtPhoneNumber.Text = "";
                    txtMessage.Text = "";

                }
                else MessageBox.Show("An error occured during sending message. Please check EJC_SMS Directory (debug.txt or error.txt) for more info.", "An error occured", MessageBoxButtons.OK, MessageBoxIcon.Information);
            }
            else // Multiple SMS
            {
                if (lboxPhoneNumbers.Items.Count < 1)
                {
                    MessageBox.Show("You haven't add any number of recipients in the list of phone numbers", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    txtAddingNumber.Focus();
                    return;
                }
                if(MessageBox.Show("Are you sure do you want to send it to multiple recipient?", "Sending Multiple Recipient", MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.Yes)
                {
                    List<string> recipients = new List<string>();
                    int countFail = 0; // Store the number of fail to send

                    // Loop all the phone numbers in lboxPhoneNumbers
                    foreach (string phoneNum in lboxPhoneNumbers.Items) recipients.Add(phoneNum);  // Add it on List<recipients>

                    // Now send the message
                    countFail = SMS.SendMultipleSMS(recipients, txtMessage.Text); 
                    string tempStr = string.Format("Number of fail to send ({0}).", countFail); // There are some error occured during sending the message
                    if (countFail == 0) tempStr = string.Format("All message has send successfully!"); // If countFail = 0 then all send with no issue
                    MessageBox.Show(tempStr, "Sending Multiple Recipient", MessageBoxButtons.OK, MessageBoxIcon.Information); // Show the message

                    // Clear this shit
                    checkSendMultiple.Checked = false;
                    lboxPhoneNumbers.Items.Clear();
                    txtMessage.Text = "";
                }
            }
        }

        private void bttnAddPhoneNumber_Click(object sender, EventArgs e)
        {
            if(!SMS.IsValidRecipientNumber(txtAddingNumber.Text))
            {
                MessageBox.Show("Invalid Phone Number!", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                txtAddingNumber.Focus();
                return;
            }
            lboxPhoneNumbers.Items.Add(txtAddingNumber.Text);
            txtAddingNumber.Text = "";
            txtAddingNumber.Focus();
        }

        private void txtAddingNumber_KeyPress(object sender, KeyPressEventArgs e)
        {
            e.Handled = !char.IsDigit(e.KeyChar) && !char.IsControl(e.KeyChar); // Allow only Numeric Key.
            if(e.KeyChar == (char)Keys.Return) bttnAddPhoneNumber.PerformClick(); // When Press Enter Perform Add Phone Number Click
        }

        private void txtPhoneNumber_KeyPress(object sender, KeyPressEventArgs e)
        {
            e.Handled = !char.IsDigit(e.KeyChar) && !char.IsControl(e.KeyChar); // Allow only Numeric Key.
            if (e.KeyChar == (char)Keys.Return) txtMessage.Focus(); // When Press enter Focus Message Box
        }

        private void bttnClear_Click(object sender, EventArgs e)
        {
            if(MessageBox.Show("Are you sure do you want to clear the list of phone numbers?", "Clearing the list of phone numbers", MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.Yes)
            {
                lboxPhoneNumbers.Items.Clear();
                checkSendMultiple.Checked = false;
            }
        }

        private void checkSendMultiple_CheckedChanged(object sender, EventArgs e)
        {
            if(checkSendMultiple.CheckState == CheckState.Checked)
            {
                if(lboxPhoneNumbers.Items.Count < 1)
                {
                    MessageBox.Show("You haven't add any number of recipients in the list of phone numbers", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    txtAddingNumber.Focus();
                    checkSendMultiple.Checked = false;
                    return;
                }
            }
        }

        private void InboxHeartbeat_Tick(object sender, EventArgs e)
        {
            // Easy Peasy showing all the SMS from your Sim Card Inbox in just one line of code.
            gridViewInbox.DataSource = SMS.ReadAllMessages();
        }
    }
}

File: EJC SMS (VBNet Version)/frmMain.vb
'
'                $$$$$$$$\   $$$$$\  $$$$$$\         $$$$$$\  $$\      $$\  $$$$$$\  
'                $$  _____|  \__$$ |$$  __$$\       $$  __$$\ $$$\    $$$ |$$  __$$\ 
'                $$ |           $$ |$$ /  \__|      $$ /  \__|$$$$\  $$$$ |$$ /  \__|
'                $$$$$\         $$ |$$ |            \$$$$$$\  $$\$$\$$ $$ |\$$$$$$\  
'                $$  __|  $$\   $$ |$$ |             \____$$\ $$ \$$$  $$ | \____$$\ 
'                $$ |     $$ |  $$ |$$ |  $$\       $$\   $$ |$$ |\$  /$$ |$$\   $$ |
'                $$$$$$$$\\$$$$$$  |\$$$$$$  |      \$$$$$$  |$$ | \_/ $$ |\$$$$$$  |
'                \________|\______/  \______/        \______/ \__|     \__| \______/ 

'                                EJC Short Message Service v0.5
'                             VB .Net Version (Converted from C#)

'                   Having trouble on making your own Send and Read Messages code for your 
'            Project / System using your GSM Modem Device? Say no more. EJC SMS (EJC Short 
'            Message Service) is a library that use AT Comamnd will let you to Send 
'            and Read Messages at ease in which helps you to minimized your time developing 
'            your system with SMS Feature.

'            Github Repository: https://github.com/eksqtr/EJC-SMS

'            For more Information about AT Commands please refer to this site:
'                https://www.developershome.com/sms/

'                                                  Developed by: https://github.com/eksqtr
'


Public Class frmMain
    'Intantiate the EJC SMS library to SMS as Object
    Dim SMS As New EJCSMS(allow_long_msg:=True, allow_empty_sms:=False) ' Configure your EJCSMS with these parameter

    Dim GSM_MODEM_PORT As String = "COM5" ' Please change this to your existing Com Port
    Private Sub bttnSend_Click(sender As Object, e As EventArgs) Handles bttnSend.Click

        If Not checkMultiple.CheckState = CheckState.Checked Then
            ' Check if the phone number is valid
            If Not SMS.IsValidRecipientNumber(txtPhoneNumber.Text) = True Then
                MessageBox.Show("Invalid Phone Number!", "Invalid Phone Number", MessageBoxButtons.OK, MessageBoxIcon.Error)
                Return
            End If
            If String.IsNullOrEmpty(txtMessage.Text) Then
                MessageBox.Show("You cannot send an Empty Message!", "Message is Empty!", MessageBoxButtons.OK, MessageBoxIcon.Error)
                txtMessage.Select() ' now focus the txtmessage
                Return
            End If
            'Now Send the message
            If SMS.SendSMS(txtPhoneNumber.Text, txtMessage.Text) = True Then
                MessageBox.Show("Message successfuly sent!", "Delivering Message Success", MessageBoxButtons.OK, MessageBoxIcon.Information)
            Else
                MessageBox.Show("Message failed to send!", "Delivering Message Failed", MessageBoxButtons.OK, MessageBoxIcon.Error)
            End If
        Else

            If String.IsNullOrEmpty(txtMessage.Text) Then
                MessageBox.Show("You cannot send an Empty Message!", "Message is Empty!", MessageBoxButtons.OK, MessageBoxIcon.Error)
                txtMessage.Select() ' now focus the txtmessage
                Return
            End If


            Dim recipient_number As New List(Of String) ' You need to declare this as List for you to able to send multiple recipient
            For Each numero As String In lboxPhoneNumber.Items ' For Each the lboxPhoneNumber ListBox 
                recipient_number.Add(numero) ' Add it on recipient_number list
            Next
            Dim tempStr As String
            Dim countFail As Integer = SMS.SendMultipleSMS(recipient_number, txtMessage.Text) ' Now you can send it to multiple recipient by doing this

            ' Check if there are number of fail to send
            If countFail > 0 Then
                tempStr = String.Format("There are {0} number of message that has been failed to deliver.", countFail)
                MessageBox.Show(tempStr, "Delivering Message has failed!", MessageBoxButtons.OK, MessageBoxIcon.Information)

            Else ' If there were no fail then just show message normaly all message has been delivered.
                tempStr = String.Format("All mesages hsa been sent successfuly")
                MessageBox.Show(tempStr, "Message Delivery Success!", MessageBoxButtons.OK, MessageBoxIcon.Information)
            End If

            ' Clear this shit.
            lboxPhoneNumber.Items.Clear()
            txtMessage.Text = ""
            checkMultiple.Checked = False
        End If

    End Sub

    Private Sub frmMain_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        ' Initialize the EJC SMS with your existing GSM Modem Device Port
        SMS.InitSMS(GSM_MODEM_PORT)

        ' Read all the mesages and set it as DataSource to gridView
        gridViewInbox.DataSource = SMS.ReadAllMessages()
        lblDateToday.Text = Date.Now.ToString("MMMM dd, yyyy")
    End Sub

    Private Sub txtAddPhoneNumber_KeyPress(sender As Object, e As KeyPressEventArgs) Handles txtAddPhoneNumber.KeyPress
        ' Allow only Numeric Keys
        If Not Char.IsNumber(e.KeyChar) AndAlso Not Char.IsControl(e.KeyChar) Then
            e.Handled = True
        End If

        'If user press enter in txtAddPhoneNumber perform addPhoneNumber button click
        If e.KeyChar = Convert.ToChar(Keys.Return) Then
            bttnAddPhoneNumber.PerformClick()
        End If
    End Sub

    Private Sub bttnAddPhoneNumber_Click(sender As Object, e As EventArgs) Handles bttnAddPhoneNumber.Click
        ' Validation if Adding phone number in multiple recipient is valid if not then this one-way selection statement will trigger.
        If SMS.IsValidRecipientNumber(txtAddPhoneNumber.Text) = False Then
            MessageBox.Show("Your not supposed to add Invalid Number!", "Invalid Phone Number", MessageBoxButtons.OK, MessageBoxIcon.Error)
            txtAddPhoneNumber.Select() ' Focus the txtbox of adding phone number
            Return
        End If

        ' Reset this shit.
        lboxPhoneNumber.Items.Add(txtAddPhoneNumber.Text)
        txtAddPhoneNumber.Text = ""
        txtAddPhoneNumber.Select()
    End Sub

    Private Sub checkMultiple_CheckedChanged(sender As Object, e As EventArgs) Handles checkMultiple.CheckedChanged

        ' Validation check if the user add multiple recipients if not then this one-way selection statement will trigger.
        If checkMultiple.CheckState = CheckState.Checked Then
            If lboxPhoneNumber.Items.Count < 1 Then
                MessageBox.Show("There were no Phone Numbers added in the phone number lists!", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
                checkMultiple.Checked = False
                txtAddPhoneNumber.Select()
            End If
        End If
    End Sub

    Private Sub txtPhoneNumber_KeyPress(sender As Object, e As KeyPressEventArgs) Handles txtPhoneNumber.KeyPress
        ' Allow only Numeric Keys
        If Not Char.IsNumber(e.KeyChar) AndAlso Not Char.IsControl(e.KeyChar) Then
            e.Handled = True
        End If
        ' If the user press enter focus textbox txtMesssage
        If e.KeyChar = Convert.ToChar(Keys.Return) Then
            txtMessage.Select()
        End If
    End Sub

    Private Sub bttnRefresh_Click(sender As Object, e As EventArgs) Handles bttnRefresh.Click

        ' Easy peasy in EJC SMS you can load the Inbox SMS with just one line of code.
        gridViewInbox.DataSource = SMS.ReadAllMessages()
    End Sub
End Class
